engine/kafka/src/main/scala/pl/touk/nussknacker/engine/kafka/validator/TopicsExistenceValidator.scala

```diff
@@ -0,0 +1,105 @@
+package pl.touk.nussknacker.engine.kafka.validator
+
+import cats.data.Validated
+import cats.data.Validated.{Invalid, Valid}
+import com.typesafe.scalalogging.LazyLogging
+import pl.touk.nussknacker.engine.kafka.KafkaConfig
+import pl.touk.nussknacker.engine.util.cache.SingleValueCache
+
+import java.util.Properties
+import scala.concurrent.duration.{DurationInt, FiniteDuration}
+import org.apache.kafka.clients.admin.{AdminClient, DescribeClusterOptions, DescribeConfigsOptions, ListTopicsOptions}
+import org.apache.kafka.common.config.ConfigResource
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError.CustomNodeError
+
+import scala.collection.JavaConverters._
+
+trait TopicsExistenceValidator {
+
+  final def validateTopic(topic: String): Validated[TopicExistenceValidationException, String] = validateTopics(List(topic)).map(_.head)
+
+  def validateTopics(topics: List[String]): Validated[TopicExistenceValidationException, List[String]]
+}
+
+trait WithCachedTopicsExistenceValidator extends TopicsExistenceValidator {
+  private lazy val validator = new CachedTopicsExistenceValidator(kafkaConfig, config)
+  protected val kafkaConfig: KafkaConfig
+  protected val config: CachedTopicsExistenceValidatorConfig = CachedTopicsExistenceValidatorConfig(
+    autoCreateFlagFetchCacheTtl = 5 minutes, topicsFetchCacheTtl = 30 seconds, adminClientTimeout = 500 millis
+  )
+
+  final override def validateTopics(topics: List[String]): Validated[TopicExistenceValidationException, List[String]] = validator.validateTopics(topics)
+}
+
+case class CachedTopicsExistenceValidatorConfig(autoCreateFlagFetchCacheTtl: FiniteDuration, topicsFetchCacheTtl: FiniteDuration, adminClientTimeout: FiniteDuration)
+
+class CachedTopicsExistenceValidator(kafkaConfig: KafkaConfig, config: CachedTopicsExistenceValidatorConfig)
+  extends TopicsExistenceValidator with LazyLogging {
+
+  private lazy val autoCreateSettingCache = new SingleValueCache[Boolean](expireAfterAccess = None, expireAfterWrite = Some(config.autoCreateFlagFetchCacheTtl))
+  private lazy val topicListCache = new SingleValueCache[Set[String]](expireAfterAccess = None, expireAfterWrite = Some(config.topicsFetchCacheTtl))
+  private lazy val kafkaAdminClient = {
+    val properties = new Properties()
+    properties.setProperty("bootstrap.servers", kafkaConfig.kafkaAddress)
+    logger.info(kafkaConfig.kafkaAddress)
+    AdminClient.create(properties)
+  }
+
+  def validateTopics(topics: List[String]): Validated[TopicExistenceValidationException, List[String]] =
+    isAutoCreateEnabled
+      .andThen(if (_) Valid(topics) else {
+        catchedErrorToInvalid {
+          topicListCache.getOrCreate {
```

Comment:
I wonder if it would be a bad idea to always has scenario:
- try cache
- if not exist in cache, fetch fresh topics
Thanks to that approach we will have 0 latency for scenario when someone just have added topic and fast response for optimistic case when topic already exit.

Reply:
I'm not quite sure about this solution, because it will cause that cache will be totally pointless when topic is not created..
I've added it in speparate commit, though.